### PR TITLE
Fix a crash for Enum class decorated with dataclass

### LINKED ISSUE #9100: Crash with dataclass enums
### Bug description

When parsing the following ``a.py``:

```python
from dataclasses import dataclass
from enum import Enum


@dataclass
class Something(str, Enum):
	asd: str = 'sdf'
```

### Configuration

_No response_

### Command used

```shell
pylint a.py
```

### Pylint output

```python
<details open>
    <summary>
        pylint crashed with a ``AstroidError`` and with the following stacktrace:
    </summary>


Traceback (most recent call last):
  File "/somewhere/test_pylint_enum/.venv/lib64/python3.11/site-packages/pylint/lint/pylinter.py", line 788, in _lint_file
    check_astroid_module(module)
  File "/somewhere/test_pylint_enum/.venv/lib64/python3.11/site-packages/pylint/lint/pylinter.py", line 1017, in check_astroid_module
    retval = self._check_astroid_module(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/somewhere/test_pylint_enum/.venv/lib64/python3.11/site-packages/pylint/lint/pylinter.py", line 1069, in _check_astroid_module
    walker.walk(node)
  File "/somewhere/test_pylint_enum/.venv/lib64/python3.11/site-packages/pylint/utils/ast_walker.py", line 94, in walk
    self.walk(child)
  File "/somewhere/test_pylint_enum/.venv/lib64/python3.11/site-packages/pylint/utils/ast_walker.py", line 94, in walk
    self.walk(child)
  File "/somewhere/test_pylint_enum/.venv/lib64/python3.11/site-packages/pylint/utils/ast_walker.py", line 94, in walk
    self.walk(child)
  File "/somewhere/test_pylint_enum/.venv/lib64/python3.11/site-packages/pylint/utils/ast_walker.py", line 91, in walk
    callback(astroid)
  File "/somewhere/test_pylint_enum/.venv/lib64/python3.11/site-packages/pylint/checkers/base/name_checker/checker.py", line 488, in visit_assignname
    if utils.is_enum_member(node) or utils.is_assign_name_annotated_with(
       ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/somewhere/test_pylint_enum/.venv/lib64/python3.11/site-packages/pylint/checkers/utils.py", line 2284, in is_enum_member
    enum_member_objects = frame.locals.get("__members__")[0].items
                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^
TypeError: 'NoneType' object is not subscriptable

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/somewhere/test_pylint_enum/.venv/lib64/python3.11/site-packages/pylint/lint/pylinter.py", line 752, in _lint_files
    self._lint_file(fileitem, module, check_astroid_module)
  File "/somewhere/test_pylint_enum/.venv/lib64/python3.11/site-packages/pylint/lint/pylinter.py", line 790, in _lint_file
    raise astroid.AstroidError from e
astroid.exceptions.AstroidError




</details>
```

### Expected behavior

No crash. Works on 2.17.7.

### Pylint version

```shell
pylint 3.0.0
astroid 3.0.0
Python 3.11.5 (main, Aug 28 2023, 00:00:00) [GCC 13.2.1 20230728 (Red Hat 13.2.1-1)]
```

### OS / Environment

```shell
NAME="Fedora Linux"
VERSION="38 (Workstation Edition)"
```

### Additional dependencies

_No response_

### PR DESCRIPTION
**Pull request by mbyrnepr2**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |
|    | :sparkles: New feature |
|    | :hammer: Refactoring   |
|    | :scroll: Docs          |

## Description

Fix a crash when an enum class which is also decorated with a ``dataclasses.dataclass`` decorator is defined.

Closes #9100

### CODE REVIEW

**Review by Pierre-Sassoulas — COMMENTED** [review]
Maybe we could add the functional test for the false negative we create (something that should return true ?) and a todo that we'll fix in 3.1 ?

**Comment by mbyrnepr2** [comment]
Ok once I create the astroid issue I'll update the doc-string of the test class with the issue number.

File: pylint/checkers/utils.py [hunk 1]
```diff
@@ -2281,5 +2281,8 @@ def is_enum_member(node: nodes.AssignName) -> bool:
     ):
         return False
 
-    enum_member_objects = frame.locals.get("__members__")[0].items
+    try:
+        enum_member_objects = frame.locals.get("__members__")[0].items
```
Comment (DanielNoord) [inline pylint/checkers/utils.py:2285]: Shall we do a check if the `get` returns `None`? In the future I think we will ask ourselves what kind of error we are catching here whereas seeing an `is not None` check is very explicit.

**Review by DanielNoord — CHANGES_REQUESTED** [review]
Thanks for picking this up!

File: pylint/checkers/utils.py (hunk 1 above)
Comment (mbyrnepr2) [reply pylint/checkers/utils.py:2285]: I went for try/except since in the majority of cases this code is run, there should be a `__members__` present; the only case I know of where it breaks currently, is when the dataclass decorator is used.
You make a good point though - at first glance it would be a bit difficult to see where the cause of the exception is. We could simplify it:
```diff
+    members = frame.locals.get("__members__")
     try:
-        enum_member_objects = frame.locals.get("__members__")[0].items
+        enum_member_objects = members[0].items
     except TypeError:
         return False
     return node.name in [name_obj.name for value, name_obj in enum_member_objects]
```

Or your suggestion:
```diff
 
-    try:
-        enum_member_objects = frame.locals.get("__members__")[0].items
-    except TypeError:
-        return False
-    return node.name in [name_obj.name for value, name_obj in enum_member_objects]
+    members = frame.locals.get("__members__")
+    if members is not None:
+        enum_member_objects = members[0].items
+        return node.name in [name_obj.name for value, name_obj in enum_member_objects]
+    return False
```

Comment (DanielNoord) [reply pylint/checkers/utils.py:2285]: I prefer the second diff, but perhaps with an early exit for `if members is None` and then doing the rest of the checks.

We can just add a comment saying that "due to unforeseen circumstances __members__ is not always present" 😄

File: pylint/checkers/utils.py [hunk 2]
```diff
@@ -2281,5 +2281,9 @@ def is_enum_member(node: nodes.AssignName) -> bool:
     ):
         return False
 
-    enum_member_objects = frame.locals.get("__members__")[0].items
+    members = frame.locals.get("__members__")
+    # A dataclass is one known case for when `members` can be `None`
+    if members is None:
+        return False
+    enum_member_objects = members[0].items
     return node.name in [name_obj.name for value, name_obj in enum_member_objects]
```
Comment (DanielNoord) [inline pylint/checkers/utils.py:2289]:
```suggestion
    return node.name in [name_obj.name for value, name_obj in members[0].items]
```

One final nit

**Review by DanielNoord — APPROVED** [review]

**Review by Pierre-Sassoulas — APPROVED** [review]
Great PR !
